server/image_pull.go

```diff
@@ -273,7 +273,7 @@ func (s *Server) pullImageCandidate(ctx context.Context, sourceCtx *imageTypes.S
 func consumeImagePullProgress(ctx context.Context, cancel context.CancelFunc, progress <-chan imageTypes.ProgressProperties, remoteCandidateName storage.RegistryImageReference) {
 	// The progress interval is 1s, but we give it a bit more time just in case
 	// that the connection revives.
-	const timeout = 10 * time.Second
+	const timeout = 60 * time.Second
```

Comment:
Hum, I didn’t realize this exists.

This is currently a mismatch with the c/image implementation; it only reports progress on “layer copies”, but a pull is not just layer copies.

In particular, in the relevant case of pulling to c/storage, the part that reports progress is just a network download, storing to a local file (and decompressing to obtain a digest); and afterwards a “commit” phase happens which actually extracts the local files into individual files for the overlay filesystem; _and the commit phase makes no progress updates at all_.

Now, the code _tries_ to interleave commit operations and starts doing them as soon as the layers (in order) finish downloading, but that’s not guaranteed to happen in general, and even if it _did_ happen perfectly, if the last layer is large, that is going to take a long time without any progress updates.

I’m not saying that it’s _good_ that it works this way; just that it works that way now, and imposing a timeout on operations which don’t report progress will inevitably cause pull errors on large enough images with slow enough storage.

Reply:
Would there be a better way of doing this?

Related:

- https://github.com/cri-o/cri-o/pull/7887

@saschagrunert was aiming to address an upstream Kubernetes issue/feature here.

Reply:
Also, this is the gist of why I increased this timeout to 60 seconds to get things working with composefs:

_(to quote from @mtrmac's comment above)_

> (...) I’m not saying that it’s good that it works this way; just that it works that way now, and imposing a timeout on operations which don’t report progress will inevitably cause pull errors on large enough images with slow enough storage.

Reply:
Right now c/image just doesn’t have a “watchdog keep-alive” feature.

Reply:
As an interim solution, I will turn this into a configuration option, so that people can adjust this and set as needed, should it be causing issues.